### PR TITLE
Add routes for get, list, and update platforms

### PR DESCRIPTION
**Pull request by tkyi**
- list
- update
- save
- get

*also added temporarily: `id` generation is current time

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * @param  {Hapi}     server            Hapi Server
+ * @param  {Object}   options           Configuration
+ * @param  {Object}   options.datastore Datastore Model
+ * @param  {Function} next              Function to call when done
+ */
+exports.register = (server, options, next) => {
+    server.route([
+        listRoute(options.datastore),
+        getRoute(options.datastore),
+        saveRoute(options.datastore)
```
Comment (d2lam) [inline index.js]: Should have a separate route for create and a separate route for update

File: lib/save.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+                    id: request.params.id,
+                    data: request.payload.data
+                }
+            };
+
+            datastore.save(config, reply);
+        },
+        validate: {
+            payload: {
+                data: Joi.object().required()
```
Comment (d2lam) [inline lib/save.js]: should use the data model to validate

File: README.md [hunk 3]
```diff
@@ -8,6 +8,52 @@
...
+`PUT /platforms/{id}`
+
+Example payload:
+```json
+{
+    "config": {},
+    "author": "monkey",
+    "docUrl": "http://blah.com",
+    "scmUrl_sha": "git@github.com:screwdriver-cd/data-model.git#c55494cdc55063ecd854800cc4ccc949659c1519",
+    "experimental": false
```
Comment (d2lam) [inline README.md]: For create: missing name, version 
scmUrl instead of scmUrl_sha

For update: only experimental

File: lib/save.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    config: {
+        description: 'Save a platform',
+        notes: 'Save a specific platform',
+        tags: ['api', 'platforms'],
+        handler: (request, reply) => {
+            const config = {
+                table: 'platforms',
+                params: {
+                    id: request.params.id,
+                    data: request.payload.data
```
Comment (d2lam) [inline lib/save.js]: request.payload

File: index.js (hunk 1 above)
Comment (d2lam) [inline index.js]: Or you can use the same route but need to take care of the difference bw create and update

File: lib/list.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+const DataModel = require('screwdriver-data-model');
+
+const ListModel = Joi.array().items(DataModel.platform.get);
+
+module.exports = (datastore) => ({
+    method: 'GET',
+    path: '/platforms',
+    config: {
+        description: 'Get platforms with pagination',
+        notes: 'Returns all platform records',
```
Comment (Filbird) [inline lib/list.js]: Return platform records with pagination.

File: lib/list.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        },
+        response: {
+            schema: ListModel
+        },
+        validate: {
+            query: {
+                page: Joi.number().integer().min(1)
+                    .default(1),
+                count: Joi.number().integer().min(1)
+                .max(50)
```
Comment (d2lam) [inline lib/list.js]: indentation

File: lib/save.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        handler: (request, reply) => {
+            const config = {
+                table: 'platforms',
+                params: {
+                    id: request.params.id,
+                    data: request.payload
+                }
+            };
+
+            console.log('meow: ', config);
```
Comment (nkatzman) [inline lib/save.js]: meow that's this doing here?

File: README.md [hunk 8]
```diff
@@ -8,6 +8,48 @@
...
+`page` and `count` are optional
+
+`GET /platforms?page={page}&count={count}`
+
+#### Returns a single platform
+`GET /platforms/{id}`
+
+#### Updates a single platform
+`PUT /platforms/{id}`
+
```
Comment (nkatzman) [inline README.md]: Should update this with the POST route

File: lib/get.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        tags: ['api', 'platforms'],
+        handler: (request, reply) => {
+            const config = {
+                table: 'platforms',
+                params: {
+                    id: request.params.id
+                }
+            };
+
+            datastore.get(config, reply);
```
Comment (nkatzman) [inline lib/get.js]: does reply(null) automagically convert to a 404 error?

File: lib/save.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+module.exports = (datastore) => ({
+    method: 'POST',
+    path: '/platforms',
+    config: {
+        description: 'Save a platform',
+        notes: 'Save a specific platform',
+        tags: ['api', 'platforms'],
+        handler: (request, reply) => {
+            const tempId = [request.payload.name, ':', request.payload.version].join('');
```
Comment (FenrirUnbound) [inline lib/save.js]: Currently does not conform to the `data-model`: https://github.com/screwdriver-cd/data-model/blob/master/model/platform.js#L7

File: test/index.test.js [hunk 11]
```diff
@@ -1,8 +1,193 @@
...
+            datastoreMock.get.withArgs(config).yieldsAsync(null, testPlatform);
+            server.inject('/platforms/1', (reply) => {
+                assert.equal(reply.statusCode, 200);
+                assert.deepEqual(reply.result, testPlatform);
+                done();
+            });
+        });
+    });
+
+    describe('/platforms/{id}', () => {
```
Comment (FenrirUnbound) [inline test/index.test.js]: Redundant `describe` block as [Line 61](https://github.com/screwdriver-cd/plugin-platforms/pull/1/files#diff-0fd0e07cf6d02bf7cf00f18cebb8e6eaR61)

File: test/index.test.js [hunk 12]
```diff
@@ -1,8 +1,193 @@
...
+                    config: {},
+                    author: 'batman',
+                    scmUrl: 'git@github.com:screwdriver-cd/data-model.git',
+                    docUrl: 'http://blah.com',
+                    experimental: false
+                }
+            };
+
+            server.inject(options, (reply) => {
+                assert.equal(reply.statusCode, 200);
```
Comment (FenrirUnbound) [inline test/index.test.js]: Should the `create` operation return a 200 or 201?

Comment (FenrirUnbound) [inline test/index.test.js]: Addressed in [Line 217](https://github.com/screwdriver-cd/plugin-platforms/pull/1/files#diff-0fd0e07cf6d02bf7cf00f18cebb8e6eaR217) 👍

**Comment by FenrirUnbound** [comment]
👍 

Looks good to me

insignificantforthrightinvisiblerail-size_restricted

**Comment by nkatzman** [comment]
👍  LGTM

File: README.md [hunk 13]
```diff
@@ -8,6 +8,66 @@
...
+Example payload:
+```json
+{
+    "experimental": false
+}
+```
+
+#### Creates/Saves a single platform
```
Comment (d2lam) [inline README.md]: Just "Creates" is fine

**Comment by d2lam** [comment]
LGTM 👍

**Comment by FenrirUnbound** [comment]
👍

**Pull request merged**